### DISCUSSION TITLE
500 - TypeError: Cannot destructure property 'nextauth' of 'req.query' as it is undefined.

### DISCUSSION THREAD

**Discussion by gavinsykes** [nextauthjs/next-auth#6798] [Help]
### Environment

```
System:
    OS: macOS 12.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 690.83 MB / 16.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 19.2.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.19.3 - /usr/local/bin/npm
  Browsers:
    Chrome: 110.0.5481.100
    Firefox: 109.0.1
    Safari: 15.2
```

### Reproduction URL

http://localhost:8010

### Describe the issue

Hi everyone, after much debugging and wrestling with both Next and NextAuth, as well as their types, I've reached an impasse where no matter what I try I juat cannot seem to get my authentication flow working.

I have NextAuth using a CredentialsProvider that polls an API for an access token, this all works fine and I can even log the access token when I log in, however my application just seemingly refuses to recognise that there is a session in place when I have logged in and will just constantly redirect me back to the login page (which is the desired functionality when I *haven't* logged in).

The closest I can get to working it out is that the context given to `getServerSideProps` has a `req` and `res`, but the `req` doesn't have a `query` property (although sometimes it logs as `{}`), which `getServerSession` needs to destructure internally - how am I meant to bridge that gap?

What I've tried:

- Just use `useSession` on the client-side instead, not ideal but if I can get that working it will do for now. Of course, it isn't working, it just gives me the "a client-side error has occurred" message with browser console logging that for all it's worth may as well be in Latin.
- Replace `getServerSession` with `getSession`, this stops throwing the 500 but will not persist the session when I log in. The flow goes:
  - Home
  - no session
  - redirect to login
  - enter details
  - login successful and token returned from API, even then polling the same API to get the user's permissions
  - callback URL in signIn function takes me back to Home
  - no session
- Use the arrow function-version of the export in `pages/api/auth/[...nextAuth].ts`
  - Both `getServerSession` and `getSession` have been tried with this as both `export default NextAuth(options);` and `export default (request: NextApiRequest, response: NextApiResponse<any>) => NextAuth(request, response, options);`

I've already downgraded and locked my version of NextAuth to `4.15.1` as that is what I've seen doesn't introduce breaking changes to accommodate Next 13 (the version of Node on our servers is too old for Next 13), in fact it's quite difficult to think of anything that I haven't tried, from any corner of the Internet that I haven't frequented, in order to get this to work.

Is there anyone at all that can help with this?

### How to reproduce

Dependencies from `package.json`

```json
"dependencies": {
    "@fortawesome/fontawesome-svg-core": "^6.1.1",
    "@fortawesome/free-regular-svg-icons": "^6.1.1",
    "@fortawesome/free-solid-svg-icons": "^6.1.1",
    "@fortawesome/react-fontawesome": "^0.1.18",
    "@tanstack/react-query": "^4.24.4",
    "axios": "^1.3.2",
    "d3-geo": "^3.0.1",
    "next": "^12",
    "next-auth": "4.15.1",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "swr": "^2.0.3",
    "zod": "^3.20.2"
  },
```

In my `pages/index.tsx` I have the below `getServerSideProps` function:

```typescript
export const getServerSideProps: GetServerSideProps = async (context) => {
  console.log('getServerSideProps called for Home');
  console.log(context);
  const session = await getServerSession(context.req, context.res as NextApiResponse, authOptions);
  console.log(`If you see this, the error wasn't on getServerSession`);

  if (!session) {
    console.log('No session:', session);
    return {
      redirect: {
        destination: '/login',
        permanent: false
      },
      props: {}
    }
  }

  return {
    props: {
      session
    }
  }
}
```

And here is my `[...nextAuth].ts` file:

```typescript
import { NextApiRequest, NextApiResponse } from "next";
import NextAuth, { NextAuthOptions, Session, User } from "next-auth";
import { JWT } from "next-auth/jwt";
import CredentialsProvider from 'next-auth/providers/credentials';

const credentialsProviderOptions = {
  credentials: {
    username: { label: "Username" , type: "text" },
    password: { label: "Password", type: "password" }
  },
  id: 'credentials',
  name: 'Credentials',
  type: "credentials" as "credentials",
  async authorize (credentials: any) {
    console.log('Authorising...');
    const { username, password } = credentials as Record<'username' | 'password' , string>;
    const authHeader = Buffer.from(`${process.env.API_CLIENT_ID}:${process.env.API_CLIENT_SECRET}`).toString('base64');
    const result = await fetch('https://url.for.api/authentication/access-token',{
      body: JSON.stringify(
        { username, password, 'grant_type': 'password' }
      ),
      headers: {
        accept: '*/*',
        Authorization: `Basic ${authHeader}`,
        'Content-Type': 'application/json'
      },
      method: 'POST'
    });
    console.log('Result: ', result);
    const user = await result.json();
    if (user) {
      console.log('User returned');
      console.log(user);
      return user
    }
    return null;
  },
};

const providers = [
  CredentialsProvider(credentialsProviderOptions)
];
const callbacks = {
  async jwt({ token, user }: { token: JWT, user?: User }) {
    console.log('jwt called');
    if (user) {
      console.log('User: ', user);
      token.accessToken = user.access_token;
    }

    console.log('Token: ', token);
    return token;
  },
  async session({ session, token }: { session: Session, token: JWT }) {
    session.user.accessToken = token.accessToken;
    const payload = (token.accessToken as string).split('.')[1];
    const decodedPayload = new Buffer(payload, 'base64').toString();
    console.log(JSON.parse(decodedPayload));
    session.user.roles = JSON.parse(decodedPayload).user_roles;
    const res = await fetch('https://url.for.api/authorisation/get-my-permissions',{
      method: 'GET',
      headers: {
        Authorization: `Bearer ${token.accessToken}`
      }
    });
    const permissions = await res.json();
    session.user.permissions = permissions.data;
    console.log(session);
    return session;
  }
};
const debug = true;
const pages = {
  error: '/error',
  signIn: '/login'
};
const session = {
  maxAge: 3540,
  strategy: 'jwt'
};
const secret = process.env.OMP_API_CLIENT_SECRET;

export const options = {
  callbacks,
  debug,
  pages,
  providers,
  secret,
  session
} as NextAuthOptions;

export default NextAuth(options);
```

### Expected behavior

Tell me that there's no session on initial load, redirect me to the login page, then persist the session when I log in.

**Comment by balazsorban44** [comment]
The file needs to be called `[...nextauth].ts`, not `[...nextAuth].ts`. You can also try upgrading to `next-auth@latest` first. Hard to tell more without a reproduction.

Some other notes:

- `secret` should be a unique value, not the same as `OMP_API_CLIENT_SECRET`.
But instead of defining it in code, you can just add `NEXTAUTH_SECRET`, that we pick up automatically. See https://next-auth.js.org/configuration/options#nextauth_secret

- I would recommend defining the auth options as a single object, to be able to add proper types via `NextAuthOptions`


UPDATE:

My apologies, I might have deemed this a non-issue prematurely. (Haven't confirmed any bugs yet though). I see the detailed description and the reason you locked in on an earlier version. I still think my 2 other points are valid, but I'll take another look at this! :pray:

**Reply by gavinsykes** [reply]
Thanks for taking a look @balazsorban44 and for the update! One avenue I'm just starting down which looks promising (of course further tests are needed) is using `await getServerSession(authOptions);` instead of `await getServerSession(context.req, context.res as NextApiResponse, authOptions);`.

Your point about the secret is indeed valid, I'd even already noted it down but never actioned it!

Regarding the filename it's I who owe you an apology there, it is written as `[...nextauth].ts` in my codebase and in my post I typed it rather than copy-pasting it, and in doing so I just naturally went to camelCase.

I have since slightly tweaked the declaration of `options` to the below; I hate having to do `as` but sometimes it's all I can do to get TypeScript to stop shouting at me, so when it stopped doing that I jumped at the opportunity to do it (IMO) properly.

```typescript
export const options: NextAuthOptions = {
  callbacks,
  debug,
  pages,
  providers,
  secret,
  session
};
```

**Comment by gavinsykes** [comment]
I don't know if this is part of the same issue @balazsorban44 but I've tried just using the useSession hook client-side for now, and I'm in a situation where it works on my laptop but not on the servers. I have painstakingly matched up the versions of Node and NPM so the build is no different. It's the same API and endpoint being called from both locations and checking the API logs it is generating and returning the access tokens no problem.

On my laptop it logs the `session` object with `user` and `expires` properties which is all good, and the `status` property of what is expected at the time. However, on the server, the `session` object is persistently `{ status: 401, message: "Unauthorized" }` (2 properties that I can't find at all in the type declarations) and, perhaps more alarmingly, `status` is always "authenticated"!

**Comment by gavinsykes** [comment]
Hey @balazsorban44 I don't suppose you were able to make any headway with this one? My deployment deadline looms ever closer and I'm in a bit of a hole with this!

**Comment by gavinsykes** [comment]
I bring good news @balazsorban44, it's been solved! Turns out the problem wasn't actually anything to do with the code so my apologies for that. The load balancer on our servers was set to send off anything with `/api` after the TLD to the API servers and not the web servers, meaning NextAuth's API requests were never reaching where they should! I couldn't tell you if I was even doing anything wrong in the first place, or if the version I originally had installed before 4.15.1 would have still worked despite us being unable to use Next13, but in the scramble to find a solution before discovering the load balancer config I did end up downgrading further, to 4.3.3.
